pkg/spoke/addon/configuration.go

```diff
@@ -0,0 +1,159 @@
+package addon
+
+import (
+	"crypto/sha256"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	addonv1alpha1 "github.com/open-cluster-management/api/addon/v1alpha1"
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	installationNamespaceAnnotation = "installNamespace"
+	registrationAnnotation          = "registrations"
+)
+
+// addOnConfig contains meta information of an addon
+type addOnConfig struct {
+	AddOnName             string
+	InstallationNamespace string
+}
+
+// registrationConfig contains necessary information for addon registration
+type registrationConfig struct {
+	AddOnName             string      `json:"addOnName,omitempty"`
+	InstallationNamespace string      `json:"installNamespace,omitempty"`
+	SignerName            string      `json:"signerName,omitempty"`
+	Subject               certSubject `json:"subject,omitempty"`
+
+	secretName string
+	hash       string
+}
+
+func (c *registrationConfig) x509Subject(clusterName, agentName string) *pkix.Name {
+	subject := &pkix.Name{
+		CommonName:         c.Subject.CommonName,
+		Organization:       c.Subject.Organization,
+		OrganizationalUnit: c.Subject.OrganizationalUnit,
+	}
+
+	// set the default common name
+	if len(subject.CommonName) == 0 {
+		subject.CommonName = defaultCommonName(clusterName, agentName, c.AddOnName)
+	}
+
+	// set the default organization if signer is KubeAPIServerClientSignerName
+	if c.SignerName == certificatesv1beta1.KubeAPIServerClientSignerName && len(subject.Organization) == 0 {
+		subject.Organization = []string{defaultOrganization(clusterName, c.AddOnName)}
+	}
+
+	return subject
+}
+
+// certSubject contains fields which are used to generate csr subject and can be customized by users
+type certSubject struct {
+	CommonName         string   `json:"commonName,omitempty"`
+	Organization       []string `json:"organization,omitempty"`
+	OrganizationalUnit []string `json:"organizationalUnit,omitempty"`
+}
+
+// getAddOnConfig returns addon configuration information from addon annotations.
+func getAddOnConfig(addOn *addonv1alpha1.ManagedClusterAddOn) (*addOnConfig, error) {
+	installationNamespace, ok := addOn.Annotations[installationNamespaceAnnotation]
+	if !ok {
+		return nil, fmt.Errorf("failed to get addon installation namesapce from addon %q", addOn.Name)
+	}
+
+	return &addOnConfig{
+		AddOnName:             addOn.Name,
+		InstallationNamespace: installationNamespace,
+	}, nil
+}
+
+// getRegistrationConfigs returns a slice of registrationConfig from annotations of a addon
+func getRegistrationConfigs(addOnName string, annotations map[string]string) ([]registrationConfig, error) {
+	ns := annotations[installationNamespaceAnnotation]
```

Comment:
if we're relying on these like an API, why are they not API fields.

Reply:
We will add them into api in another pr. https://github.com/open-cluster-management/api/pull/73